### PR TITLE
Refuse min proto version > max proto version

### PR DESCRIPTION
**Pull request by tiran**
SSL_CTX_set_min_proto_version() and SSL_CTX_set_max_proto_version() as
well as their SSL equivalents now validate that the mininum protocol
version is lower or requal the maximum protocol version.

Add test cases for min/max protocol API

Closes: https://github.com/openssl/openssl/issues/5127
Signed-off-by: Christian Heimes <christian@python.org>

Also see https://github.com/openssl/openssl/pull/5126

### CODE REVIEW

**Comment by kroeckx** [comment]
I think checking max > min was intentionally not done at the
moment of setting the variables. The problem is that depending on
the old values you might need to change min before max, or the other
way around.

File: test/ssl_ctx_test.c [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+};
+
+static int test_set_min_max_version(int idx_tst)
+{
+    SSL_CTX *ctx = NULL;
+    SSL *ssl = NULL;
+    int testresult = 0;
+    version_test t = version_testdata[idx_tst];
+
+    ctx = SSL_CTX_new(TLS_server_method());
```
Comment (richsalz) [inline test/ssl_ctx_test.c]: Can you use TEST_PTR for these?

File: test/ssl_ctx_test.c [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    if (ctx == NULL)
+        goto end;
+
+    ssl = SSL_new(ctx);
+    if (ssl == NULL)
+        goto end;
+
+    if (!TEST_int_eq(SSL_CTX_set_min_proto_version(ctx, t.min_version), t.min_ok))
+        goto end;
+    if (!TEST_int_eq(SSL_CTX_set_max_proto_version(ctx, t.max_version), t.max_ok))
```
Comment (richsalz) [inline test/ssl_ctx_test.c]: We tend to cascade all the if tests together for this kind of thing.

**Comment by richsalz** [comment]
ooops, meant to check 'reqeust changes'

File: test/ssl_ctx_test.c (hunk 2 above)
Comment (tiran) [reply test/ssl_ctx_test.c]: Like this? It's still short circuit evaluation.

```
    if (!TEST_int_eq(SSL_CTX_set_min_proto_version(ctx, t.min_version), t.min_ok) ||
        !TEST_int_eq(SSL_CTX_set_max_proto_version(ctx, t.max_version), t.max_ok) ||
        !TEST_int_eq(SSL_CTX_get_min_proto_version(ctx), t.expected_min) ||
        !TEST_int_eq(SSL_CTX_get_max_proto_version(ctx), t.expected_max)
       )
        goto end;
```

Comment (richsalz) [reply test/ssl_ctx_test.c]: yes like that, except put the || at the start of the line and double-indent the tests and move up the close paren.  See enctest.c, etc., for examples.

**Comment by davidben** [comment]
+1 to what @kroeckx said. The problem with this sort of thing is that it causes an interdependence between two config options, which is really messy. (See the complexities around configuring private key and certificate and when you check for consistency.)

For something like this where the check doesn't really do much useful (the handshake will just fail immediately), it's much simpler to defer to that rather than add logic everywhere.

**Comment by davidben** [comment]
For instance, consider this:

```
void ConfigureMinAndMaxVersion(SSL *ssl, uint16_t min_version, uint16_t max_version) {
#if 1
  SSL_set_min_proto_version(ssl, min_version);
  SSL_set_max_proto_version(ssl, max_version);
#else
  SSL_set_max_proto_version(ssl, max_version);
  SSL_set_min_proto_version(ssl, min_version);
#endif
}

void ChangeMyMind(SSL *ssl) {
  ConfigureMinAndMaxVersion(ssl, TLS1_1_VERSION, TLS1_1_VERSION);  // 1
  ConfigureMinAndMaxVersion(ssl, TLS1_2_VERSION, TLS1_2_VERSION);  // 2
  ConfigureMinAndMaxVersion(ssl, TLS1_VERSION, TLS1_VERSION);  // 3
}
```

Before, either version would work fine. Versions may be temporarily out of sync and fixed later. With this PR, setting the min version first doesn't work because going from 1 to 2 breaks. Setting the max version first doesn't work because going from 2 to 3 breaks.

**Comment by kaduk** [comment]
This is similar to the situation in #5057 and #5065, where the configuration APIs can be applied in different orders, and as such is too soon to effectively apply sanity checks to the configured values.  As such, I think this PR should be rejected.

**Comment by kaduk** [comment]
Whoops, I put some links in commits to this pull request that were supposed to go to #5126.
So, for future reference...

**Comment by kroeckx** [comment]
So I think we should just keep the test suite changes?

**Comment by tiran** [comment]
@kroeckx Sounds reasonable to me.

I'm currently at a conference. I'll attend to this PR and my other PRs either next week or the week after.

**Comment by kaduk** [comment]
@tiran it sounds like we are waiting for an update to trim this down to just  the tes suite changes, does that match your understanding?

**Comment by tiran** [comment]
Yes.

I'm sorry that I haven't had time to rework my patches. I spent all my free OSS time to get Python 3.7 improvement done before final beta freeze.

**Comment by kaduk** [comment]
Understandable.  (But note that we've only got two weeks until the openssl 1.1.1 feature freeze with the current schedule, https://www.openssl.org/policies/releasestrat.html)

**Comment by mattcaswell** [comment]
Ping @tiran

**Comment by levitte** [comment]
This needs a rebase, and code needs to be corrected (see the CI results)

**Comment by levitte** [comment]
Actually, looking more closely at the CI issues, I figure that a rebase will fix them as well.

**Comment by levitte** [comment]
I think we lost @tiran.  I can take this over like I've done with others...  good idea?

**Comment by richsalz** [comment]
Please do take it over!  I gave it a try and failed.

**Comment by levitte** [comment]
Continued in #6553, closing this

**Pull request closed without merging**